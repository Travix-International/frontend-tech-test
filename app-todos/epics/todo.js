import { Observable } from 'rxjs/Observable';
import { mergeMap } from 'rxjs/operators/mergeMap';
import { filter } from 'rxjs/operators/filter';

import {
    TODOS_FETCH,
    TODOS_FETCH_ASYNC,
    TODOS_ADD,
    TODOS_ADD_ASYNC,
    TODOS_FAILED
} from '../constants';

//TODO: Change to enviroment variable
const urlApi = "http://localhost:9001";

export function fetchTodosAsync$(action$) {
    return action$
        .pipe(
            filter(action => action.type === TODOS_FETCH_ASYNC),
            mergeMap(() => {
                return Observable.ajax.getJSON(`${urlApi}/tasks`)
                    .map(data => ({
                        type: TODOS_FETCH,
                        response: data.tasks,
                    }))
                    .catch(error => [
                        {
                            type: TODOS_FAILED,
                            response: { message: error.message, status: error.status },
                            error: true,
                        },
                    ])
            })
        )
}

export function addTodoAsync$(action$) {
    return action$
        .pipe(
            filter(action => action.type === TODOS_ADD_ASYNC),
            mergeMap((todo) => {
                return Observable.ajax.post(`${urlApi}/task/create/${todo.title}/${todo.description}`)
                    .map(data => ({
                        type: TODOS_ADD,
                        todo: data.response.task
                    }))
                    .catch(error => [
                        {
                            type: TODOS_FAILED,
                            response: { message: error.message, status: error.status },
                            error: true,
                        },
                    ])
            })
        )
}