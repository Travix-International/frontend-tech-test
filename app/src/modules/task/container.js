import { connect } from 'react-redux';
import component from './component';
import {
  taskList, taskDelete
} from './actions';

const mapStateToProps = state => ({
  main: state.main,
  tasks: state.main.tasks,
  isFetching: state.main.isFetching,
  notification: state.main.notification
});

const mapDispatchToProps = ({
  taskList,
  taskDelete
});

export default connect(mapStateToProps, mapDispatchToProps)(component);
