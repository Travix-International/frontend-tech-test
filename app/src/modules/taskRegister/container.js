import { connect } from 'react-redux';
import component from './component';
import {
  taskSave
} from './actions';

const mapStateToProps = (state, ownProps) => {
  const task = state.main.tasks[ownProps.match.params.id];
  return {
    task,
    isFetching: state.main.isFetching
  };
};

const mapDispatchToProps = ({
  taskSave
});

export default connect(mapStateToProps, mapDispatchToProps)(component);
