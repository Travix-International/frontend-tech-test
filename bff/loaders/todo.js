import DataLoader from 'dataloader';
import fetch from 'node-fetch';
import { restApi as restApiConfig } from "../../config.json";

const BASE_URL = `http://localhost:${restApiConfig.port}`;

const http = (path, method = "GET", body) => {
  const options = {
    method,
    headers: {
      'Content-Type': 'application/json'
    }
  }

  if (body) {
    options.body = JSON.stringify(body);
  }

  return fetch(`${BASE_URL}/${path}`, options)
    .then(res => {
      if(res.status !== 204) {
        return res.json();
      }

      return Promise.resolve();
    });
};

const getTodos = () => http("task"),
      getTodo = (id) => http(`task/${id}`),
      updateTodo = (args) => http("task", "PUT", args),
      createTodo = (args) => http("task", "POST", args),
      deleteTodo = (id) => http(`task/${id}`, "DELETE");

export const todo = () => {
  const cacheMap = new Map();

  const todosLoader = new DataLoader(keys => Promise.all(keys.map(getTodos)), { cacheMap });
  const todoLoader = new DataLoader(keys => Promise.all(keys.map(getTodo)), {
    cacheKeyFn: key => `/task/${key}/`,
    cacheMap,
  });

  todoLoader.loadAll = todosLoader.load.bind(todosLoader, '__all__');
  todoLoader.update = updateTodo;
  todoLoader.create = createTodo;
  todoLoader.delete = deleteTodo;

  return todoLoader;
}
