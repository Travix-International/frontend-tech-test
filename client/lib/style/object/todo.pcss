.o-todos-wrapper {
  lost-center: 980px;
}

.o-todos-container {
  lost-column: 1/1;

  @media (--medium-viewport) {
    lost-column: 3/5;
    lost-offset: 1/5;
  }

  @media (--large-viewport) {
    lost-column: 2/5;
    lost-offset: 1/4;
  }
}

.o-todo {
  margin: 16px 0;
  border-bottom: 1px solid var(--color-sea-foam-green);

  &__title {
    margin: 16px 0;
    font-size: 18px;
    font-weight: bold;
    color: var(--color-awesome);
  }

  &__description {
    margin: 16px 0;
    font-size: 16px;
    color: var(--color-lapis-lazuli);
  }

  &__edit {
    width:100%;
    background-color: color(var(--color-very-pale-yellow) shade(2));
    border:none;
    height: 22px;
    padding: 12px 0;

    &:focus {
      border-bottom: 1px dashed var(--color-sea-foam-green);
      outline: none;
    }
  }
}
