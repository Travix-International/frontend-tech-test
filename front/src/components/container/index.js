import styled from 'styled-components'

export default styled.section`
  max-width: 50rem;
  width: 100%;
  padding-right: 1.5rem;
  padding-left: 1.5rem;
  box-sizing: border-box;
  position: relative;
  overflow: none;
  transform: translateY(-1rem);
`
