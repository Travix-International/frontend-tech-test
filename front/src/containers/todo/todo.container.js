import React, { Component } from 'react'
import PropTypes from 'prop-types'
import ToDoComponent from './todo'
import Loading from '../../components/loading'
import Main from '../../components/main'
import Error from '../../components/error'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { getItems } from '../../modules/todo/actions/getItems'
import { insertOrUpdateItem } from '../../modules/todo/actions/insertOrUpdateItem'
import { deleteItem } from '../../modules/todo/actions/deleteItem'

export class ToDoContainer extends Component {

  constructor(props) {
    super(props)
  }

  componentDidMount() {
    this.props.getItems()
  }

  handleSaveItem = (item) => {
    const { insertOrUpdateItem } = this.props
    insertOrUpdateItem(item)
  }

  handleRemoveItem = (id) => {
    const { deleteItem } = this.props
    deleteItem(id)
  }

  render() {
    const { items: { data, loading, error } } = this.props
    return (
      <Main>
        <Loading isOpen={loading} />
        <Error isOpen={!!error} message={error} />
        <ToDoComponent
          handleRemoveItem={this.handleRemoveItem}
          handleSaveItem={this.handleSaveItem}
          items={data}
        />
      </Main>
    )
  }
}

ToDoContainer.propTypes = {
  getItems: PropTypes.func.isRequired,
  insertOrUpdateItem: PropTypes.func.isRequired,
  deleteItem: PropTypes.func.isRequired,
  items: PropTypes.object.isRequired
}

export const mapStateToProps = ({items}) => ({items})

export const mapDispatchToProps = dispatch => bindActionCreators({
  getItems,
  insertOrUpdateItem,
  deleteItem
}, dispatch)

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ToDoContainer)
