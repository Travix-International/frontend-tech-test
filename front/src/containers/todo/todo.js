import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import Card from '../../components/card'
import Grid from '../../components/grid'
import GridItem from '../../components/grid-item'
import Empty from '../../components/empty'
import CircleButton from '../../components/circle-button'
import Modal from '../../components/modal'
import ToDoForm from './components/todo-form'
import Button from './components/button'
import Title from './components/title'

class ToDoComponent extends Component {

  constructor(props) {
    super(props)
    this.state = {
      isOpenModal: false,
      item: {
        title: '',
        description: '',
        id: 0
      }
    }
  }

  cleanValuesItem = () => {
    this.setState({ item: { title: '', description: '', id: 0 } })
  }

  handleCloseModal = () => {
    this.setState({ isOpenModal: false})
    this.cleanValuesItem()
  }

  handleOpenModal = () => {
    this.setState({ isOpenModal: true})
  }

  handleClickEdit = (item) => {
    return () => {
      this.setState({ item })
      this.handleOpenModal()
    }
  }

  handleChangeInput = (prop) => {
    return (e) => {
      this.setState({
        item: {
          ...this.state.item,
          [prop]: e.target.value
        }
      })
    }
  }

  handleSaveItem = () => {
    const { item } = this.state
    const { handleSaveItem } = this.props
    handleSaveItem(item)
    this.handleCloseModal()
  }

  handleRemoveItem = (id) => {
    const { handleRemoveItem } = this.props
    return () => {
      handleRemoveItem(id)
    }
  }

  render() {
    const { isOpenModal, item } = this.state
    const { items } = this.props
    return (
      <Fragment>
        <Card>
          <Title>ToDo</Title>
          {items.length > 0 ? this.renderGrid() : <Empty />}
        </Card>
        <Button>
          <CircleButton onClick={this.handleOpenModal}/>
        </Button>
        <Modal onClose={this.handleCloseModal} open={isOpenModal}>
          <ToDoForm
            values={item}
            onChangeInput={this.handleChangeInput}
            onSave={this.handleSaveItem}/>
        </Modal>
      </Fragment>
    )
  }

  renderGrid() {
    const { items } = this.props
    return (<Grid>
      {
        items.map(item => <GridItem
          key={item.id}
          title={item.title}
          description={item.description}
          handleRemove={this.handleRemoveItem(item.id)}
          handleEdit={this.handleClickEdit(item)} />)
      }
    </Grid>)
  }
}

ToDoComponent.propTypes = {
  handleRemoveItem: PropTypes.func.isRequired,
  handleSaveItem: PropTypes.func.isRequired,
  items: PropTypes.array.isRequired
}

export default ToDoComponent
