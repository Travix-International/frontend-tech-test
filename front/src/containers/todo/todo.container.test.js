import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { ToDoContainer, mapStateToProps, mapDispatchToProps } from './todo.container'

let options  = {}
describe('ToDoContainerComponent', () => {
  beforeEach(() => {
    options = {
      getItems: jest.fn(),
      insertOrUpdateItem: jest.fn(),
      deleteItem: jest.fn(),
      items: {
        data: [],
        loading: false,
        error: ''
      }
    }
  })
  it('Snapshot with loading', async () => {
    options.items.loading = true
    const tree = shallow(<ToDoContainer {...options} />)
    expect(toJson(tree)).toMatchSnapshot()
  })
  it('Snapshot with error', async () => {
    options.items.error = 'Error Network'
    const tree = shallow(<ToDoContainer {...options} />)
    expect(toJson(tree)).toMatchSnapshot()
  })
  it('getItems function', async () => {
    shallow(<ToDoContainer {...options} />)
    expect(options.getItems).toHaveBeenCalled()
  })
  it('handleSaveItem', async () => {
    const item = {
      id: 'asd',
      title: 'title',
      description: 'description'
    }
    const tree = shallow(<ToDoContainer {...options} />)
    const instance = tree.instance()
    instance.handleSaveItem(item)
    expect(options.insertOrUpdateItem).toHaveBeenCalledWith(item)
  })
  it('handleRemoveItem', async () => {
    const id = 'asd'
    const tree = shallow(<ToDoContainer {...options} />)
    const instance = tree.instance()
    instance.handleRemoveItem(id)
    expect(options.deleteItem).toHaveBeenCalledWith(id)
  })
  it('execute maps methods', () => {
    mapStateToProps({
      items: {
        data: []
      }
    })
    mapDispatchToProps({})
  })
})
