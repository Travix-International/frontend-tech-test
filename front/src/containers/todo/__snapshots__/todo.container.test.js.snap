// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoContainerComponent Snapshot with error 1`] = `
<styled.section>
  <Loading
    isOpen={false}
  />
  <ErrorComponent
    isOpen={true}
    message="Error Network"
  />
  <ToDoComponent
    handleRemoveItem={[Function]}
    handleSaveItem={[Function]}
    items={Array []}
  />
</styled.section>
`;

exports[`ToDoContainerComponent Snapshot with loading 1`] = `
<styled.section>
  <Loading
    isOpen={true}
  />
  <ErrorComponent
    isOpen={false}
    message=""
  />
  <ToDoComponent
    handleRemoveItem={[Function]}
    handleSaveItem={[Function]}
    items={Array []}
  />
</styled.section>
`;
